odl/solvers/functional/derivatives.py

```diff
@@ -0,0 +1,330 @@
+            raise ValueError("`method` '{}' not understood").format(method_in)
+
+        Operator.__init__(self, operator.domain, operator.range,
+                          linear=True)
+
+    def _call(self, dx):
+        """Return ``self(x)``."""
+        x = self.point
+
+        dx_norm = dx.norm()
+        if dx_norm == 0:
+            return 0
+
+        scaled_dx = dx * (self.step / dx_norm)
+
+        if self.method == 'backward':
+            dAdx = self.operator(x) - self.operator(x - scaled_dx)
+        elif self.method == 'forward':
+            dAdx = self.operator(x + scaled_dx) - self.operator(x)
+        elif self.method == 'central':
+            dAdx = (self.operator(x + scaled_dx / 2) -
+                    self.operator(x - scaled_dx / 2))
+        else:
+            raise RuntimeError('unknown method')
+
+        return dAdx * (dx_norm / self.step)
+
+
+class NumericalGradient(Operator):
+
+    """The gradient of a `Functional` computed by finite differences.
+
+    See Also
+    --------
+    NumericalDerivative : Compute directional derivative
+    """
+
+    def __init__(self, functional, method='forward', step=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        functional : `Functional`
+            The functional whose gradient should be computed. Its domain must
+            be an `FnBase` space.
+        method : {'backward', 'forward', 'central'}
+            The method to use to compute the gradient.
+        step : float
+            The step length used in the derivative computation.
+            Default: selects the step according to the dtype of the space.
+
+        Examples
+        --------
+        >>> import odl
+        >>> space = odl.rn(3)
+        >>> func = odl.solvers.L2NormSquared(space)
+        >>> grad = NumericalGradient(func)
+        >>> grad([1, 1, 1])
+        rn(3).element([2.0, 2.0, 2.0])
+
+        The gradient gives the correct value with sufficiently small step size:
+
+        >>> grad([1, 1, 1]) == func.gradient([1, 1, 1])
+        True
+
+        If the step is too large the result is not correct:
+
+        >>> grad = NumericalGradient(func, step=0.5)
+        >>> grad([1, 1, 1])
+        rn(3).element([2.5, 2.5, 2.5])
+
+        But it can be improved by using the more accurate ``method='central'``:
+
+        >>> grad = NumericalGradient(func, method='central', step=0.5)
+        >>> grad([1, 1, 1])
+        rn(3).element([2.0, 2.0, 2.0])
+
+        Notes
+        -----
+        If the functional is :math:`f` and step size :math:`h` is used, the
+        gradient is computed as follows.
+
+        ``method='backward'``:
+
+        .. math::
+            \\frac{f(x) - f(x - h)}{h}
+
+        ``method='forward'``:
+
+        .. math::
+            \\frac{f(x + h) - f(x)}{h}
+
+        ``method='central'``:
+
+        .. math::
+            \\frac{f(x + h/2) - f(x - h/2)}{h}
```

Comment:
This is not the same as the code. What happens is (here for backward)
```
(\\nabla f(x))_i = \\frac{f(x) - f(x - h e_i)}{h}
```
for all standard unit vectors `e_i`.